ghcid-ng/src/ghci/parse/haskell_grammar.rs

```diff
@@ -0,0 +1,74 @@
+//! Parse elements of the Haskell grammar.
+//!
+//! See: ["Lexical Structure" in "The Haskell 2010 Language".][1]
+//!
+//! [1]: https://www.haskell.org/onlinereport/haskell2010/haskellch2.html
+
+use winnow::combinator::separated1;
+use winnow::token::one_of;
+use winnow::token::take_while;
+use winnow::PResult;
+use winnow::Parser;
+
+/// A Haskell module name.
+///
+/// See: `modid` in <https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-180002.4>
+pub fn module_name<'i>(input: &mut &'i str) -> PResult<&'i str> {
+    // Surely there's a better way to get type inference to work here?
+    separated1::<_, _, (), _, _, _, _>(constructor_name, ".")
+        .recognize()
+        .parse_next(input)
+}
+
+/// A Haskell constructor name.
+///
+/// See: `conid` in <https://www.haskell.org/onlinereport/haskell2010/haskellch2.html#x7-180002.4>
+fn constructor_name<'i>(input: &mut &'i str) -> PResult<&'i str> {
+    // TODO: Support Unicode letters.
```

Comment:
We can do this with the nice `unicode-properties` crate, but I'm not sure we'll need it?

https://docs.rs/unicode-properties/latest/unicode_properties/